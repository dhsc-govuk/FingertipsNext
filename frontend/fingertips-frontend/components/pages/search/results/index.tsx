'use client';

import { BackLink, H1, Paragraph } from 'govuk-react';

export function SearchResults({ indicator }: Readonly<{ indicator: string }>) {
  return (
    <>
      <BackLink href={`/search?indicator=${indicator}`} />
      <H1>Search results</H1>
      <Paragraph>{`You searched for indicator "**${indicator}**"`}</Paragraph>
    </>
  );
}
