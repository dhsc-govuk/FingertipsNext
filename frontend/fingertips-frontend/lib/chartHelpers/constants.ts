export const areaCodeForEngland = 'E92000001';
export const englandAreaString = 'England';
export const indicatorIdForPopulation = 337;
