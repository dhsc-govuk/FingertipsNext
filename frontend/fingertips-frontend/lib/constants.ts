export const siteTitle = 'Find public health data';

export const siteDescription = `${siteTitle} is a rich source of indicators across a range of health and wellbeing themes designed to support JSNA and commissioning to improve health and wellbeing, and reduce inequalities.`;

export const spaceSeparatedPattern = /^\d+(\s+\d+)+$/;
