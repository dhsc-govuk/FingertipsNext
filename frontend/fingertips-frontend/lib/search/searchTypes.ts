/*********************************************************************************
 * This file includes duplicate definitions from the search-setup/src/constants.ts
 * These should be updated to take definitions from there rather than duplicating them
 */

export const INDICATOR_SEARCH_SCORING_PROFILE = 'basicScoringProfile';
export const INDICATOR_SEARCH_INDEX_NAME = 'indicator-search-index';
export const AREA_SEARCH_INDEX_NAME = 'area-search-index';
export const AREA_SEARCH_SUGGESTER_NAME = 'areaSuggester';

export type IndicatorDocument = {
  indicatorId: string;
  name: string;
  definition: string;
  dataSource: string;
  latestDataPeriod: string;
  lastUpdated: Date;
};

export type AreaDocument = {
  areaCode: string;
  areaType: string;
  areaName: string;
};

export interface IIndicatorSearchService {
  searchWith(searchTerm: string): Promise<IndicatorDocument[]>;
}

export interface IAreaSearchService {
  getAreaSuggestions(partialAreaName: string): Promise<AreaDocument[]>;
}
