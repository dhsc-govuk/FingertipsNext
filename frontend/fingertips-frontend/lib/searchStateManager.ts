import { asArray } from './pageHelpers';

export enum SearchParams {
  SearchedIndicator = 'si',
  IndicatorsSelected = 'is',
  AreasSelected = 'as',
  AreaTypeSelected = 'ats',
  GroupTypeSelected = 'gts',
  ConfidenceIntervalSelected = 'cis',
}

export type SearchParamKeys = `${SearchParams}`;

const multiValueParams = [
  SearchParams.IndicatorsSelected as string,
  SearchParams.AreasSelected as string,
  SearchParams.ConfidenceIntervalSelected as string,
];

export type SearchStateParams = {
  [SearchParams.SearchedIndicator]?: string;
  [SearchParams.IndicatorsSelected]?: string[];
  [SearchParams.AreasSelected]?: string[];
  [SearchParams.AreaTypeSelected]?: string;
  [SearchParams.GroupTypeSelected]?: string;
  [SearchParams.ConfidenceIntervalSelected]?: string[];
};

const isMultiValueTypeParam = (searchParamKey: SearchParamKeys) =>
  multiValueParams.includes(searchParamKey);

export class SearchStateManager {
  private searchState: SearchStateParams;
  private searchStateParams: URLSearchParams;

  constructor(searchState: SearchStateParams = {}) {
    this.searchState = {};
    this.searchStateParams = new URLSearchParams();

    this.searchState = searchState;
  }

  private constructPath(path: string) {
    if (this.searchStateParams.size === 0) {
      return path;
    }
    return `${path}?${this.searchStateParams.toString()}`;
  }

  private addParamToPath(searchParamKey: SearchParamKeys) {
    const paramsToAdd = asArray(this.searchState[searchParamKey]);

    if (paramsToAdd.length > 0) {
      paramsToAdd.forEach((paramToAdd) => {
        this.searchStateParams.append(searchParamKey, paramToAdd);
      });
    }
  }

  public addParamValueToState(
    searchParamKey: SearchParamKeys,
    paramToAdd: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = this.searchState[searchParamKey];

      const currentParamStateAsArray = asArray(currentParamState);
      const doesParamExist = currentParamStateAsArray.find(
        (currentParam) => currentParam === paramToAdd
      );

      if (!doesParamExist) {
        const newState: SearchStateParams = {
          ...this.searchState,
          [searchParamKey]: [...currentParamStateAsArray, paramToAdd],
        };

        this.searchState = newState;
      }
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: paramToAdd,
      };

      this.searchState = newState;
    }
  }

  public removeParamValueFromState(
    searchParamKey: SearchParamKeys,
    paramToRemove?: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = asArray(this.searchState[searchParamKey]);

      const newParamState = currentParamState?.filter((paramValue) => {
        return paramValue !== paramToRemove;
      });

      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: newParamState,
      };

      this.searchState = newState;
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: undefined,
      };

      this.searchState = newState;
    }
  }

  public removeAllParamFromState(searchParamKey: SearchParamKeys) {
    this.searchState[searchParamKey] = undefined;
  }

  public static setStateFromParams(params: URLSearchParams) {
    const newState = Object.values(SearchParams).reduce<SearchStateParams>(
      (state, searchParamKey) => {
        if (isMultiValueTypeParam(searchParamKey)) {
          const paramValues = params.getAll(searchParamKey);

          if (paramValues.length > 0) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValues,
            };

            return newState;
          }
        } else {
          const paramValue = params.get(searchParamKey);

          if (paramValue) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValue,
            };

            return newState;
          }
        }
        return state;
      },
      {}
    );

    const searchStateManager = new SearchStateManager(newState);
    return searchStateManager;
  }

  public static setStateFromSearchStateParams(params?: SearchStateParams) {
    if (params) {
      const newState = Object.values(SearchParams).reduce<SearchStateParams>(
        (state, searchParamKey) => {
          if (isMultiValueTypeParam(searchParamKey)) {
            const paramValues = asArray(params[searchParamKey]);

            if (paramValues.length > 0) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValues,
              };

              return newState;
            }
          } else {
            const paramValue = params[searchParamKey];

            if (paramValue) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValue,
              };

              return newState;
            }
          }
          return state;
        },
        {}
      );

      const searchStateManager = new SearchStateManager(newState);
      return searchStateManager;
    }
    return new SearchStateManager();
  }

  public getSearchState() {
    return this.searchState;
  }

  public generatePath(path: string) {
    this.searchStateParams = new URLSearchParams();

    Object.values(SearchParams).forEach((searchParamKey) => {
      this.addParamToPath(searchParamKey);
    });

    return this.constructPath(path);
  }
}
