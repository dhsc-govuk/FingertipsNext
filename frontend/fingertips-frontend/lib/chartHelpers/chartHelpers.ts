import { HealthDataForArea } from '@/generated-sources/ft-api-client';
import { areaCodeForEngland } from './constants';

export function sortHealthDataByDate(
  data: HealthDataForArea[]
): HealthDataForArea[] {
  return data.map((item) => ({
    ...item,
    healthData: item.healthData.toSorted((a, b) => a.year - b.year),
  }));
}

export function sortHealthDataByYearDescending(
  data: HealthDataForArea[] = []
): HealthDataForArea[] {
  return data.map((item) => ({
    ...item,
    healthData: item.healthData.toSorted((a, b) => b.year - a.year),
  }));
}

export function seriesDataForIndicatorIndexAndArea(
  data: HealthDataForArea[][],
  indicatorIndex: number,
  seriesAreaCode: string
) {
  return data[indicatorIndex].find(
    (areaData) => areaData.areaCode === seriesAreaCode
  );
}

export function seriesDataWithoutEnglandOrParent(
  data: HealthDataForArea[],
  parentAreaCode?: string
) {
  return data.filter(
    (item) =>
      item.areaCode !== areaCodeForEngland && item.areaCode !== parentAreaCode
  );
}
