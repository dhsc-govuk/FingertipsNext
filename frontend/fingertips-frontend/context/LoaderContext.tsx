'use client';

import React, {
  ReactNode,
  createContext,
  useContext,
  useMemo,
  useState,
  useEffect,
  Suspense,
  Dispatch,
  SetStateAction,
} from 'react';
import { usePathname } from 'next/navigation';
import { LoadingBox } from 'govuk-react';
import { useSearchStateParams } from '@/components/hooks/useSearchStateParams';
import { SearchStateParams } from '@/lib/searchStateManager';

export type LoaderContext = {
  getIsLoading: () => boolean;
  setIsLoading: (loading: boolean) => void;
};

type LoaderContextProvider = {
  children: ReactNode;
};

const LoaderContext = createContext<LoaderContext | undefined>(undefined);

export const LoaderProvider: React.FC<LoaderContextProvider> = ({
  children,
}) => {
  const pathname = usePathname();
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const [isLoading, setIsLoading] = useState<boolean>(true);

  const contextValue: LoaderContext = useMemo(
    () => ({
      getIsLoading: () => isLoading,
      setIsLoading,
    }),
    [isLoading, setIsLoading]
  );

  useEffect(() => {
    setIsLoading(false);
  }, [setIsLoading, pathname, searchState]);

  return (
    <LoaderContext.Provider value={contextValue}>
      <Suspense>
        <SuspendedUseSearchStateParams setSearchState={setSearchState} />
      </Suspense>
      <LoadingBox loading={isLoading} timeIn={200} timeOut={200}>
        {children}
      </LoadingBox>
    </LoaderContext.Provider>
  );
};

export const useLoadingState = () => {
  const context = useContext(LoaderContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};

// Next has a very peculiar quirk where useSearchState cannot be called server
// side, which works fine during dev but build will fail.
// Here we're creating a component which will not render server side by
// suspending it until client mounts at which point it can set the loading
// context value, all this is more complicated because loader context is in a
// layout.
function SuspendedUseSearchStateParams({
  setSearchState,
}: Readonly<{ setSearchState: Dispatch<SetStateAction<SearchStateParams>> }>) {
  const searchParams = useSearchStateParams();
  useEffect(() => {
    setSearchState(searchParams);
  }, [searchParams, setSearchState]);

  return null;
}
