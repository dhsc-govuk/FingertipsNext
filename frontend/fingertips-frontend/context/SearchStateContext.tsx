'use client';

import React, { ReactNode, createContext, useContext, useState } from 'react';
import { SearchStateParams } from '@/lib/searchStateManager';

export type SearchStateContext = {
  getSearchState: () => SearchStateParams;
  setSearchState: (state: SearchStateParams) => void;
};

type SearchStateContextProvider = {
  children: ReactNode;
};

const SearchStateContext = createContext<SearchStateContext | undefined>(
  undefined
);

export const SearchStateProvider: React.FC<SearchStateContextProvider> = ({
  children,
}) => {
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const contextValue: SearchStateContext = {
    getSearchState: () => searchState,
    setSearchState,
  };

  return (
    <SearchStateContext.Provider value={contextValue}>
      {children}
    </SearchStateContext.Provider>
  );
};

export const useSearchState = () => {
  const context = useContext(SearchStateContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};
