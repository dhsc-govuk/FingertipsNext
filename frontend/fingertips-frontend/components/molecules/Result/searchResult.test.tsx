import { render, screen } from '@testing-library/react';
import { expect } from '@jest/globals';
import { SearchResult } from '.';
import { UserEvent, userEvent } from '@testing-library/user-event';
import { SearchParams, SearchStateParams } from '@/lib/searchStateManager';
import { IndicatorDocument } from '@/lib/search/searchTypes';
import { areaCodeForEngland } from '@/lib/chartHelpers/constants';
import { LoaderContext } from '@/context/LoaderContext';
import { SearchStateContext } from '@/context/SearchStateContext';

let user: UserEvent;

const MOCK_DATA: IndicatorDocument[] = [
  {
    indicatorID: '1',
    indicatorName: 'NHS',
    indicatorDefinition:
      'Total number of patients registered with the practice',
    earliestDataPeriod: '1999',
    latestDataPeriod: '2023',
    dataSource: 'NHS website',
    lastUpdatedDate: new Date('December 6, 2024'),
    hasInequalities: false,
    unitLabel: '',
  },
  {
    indicatorID: '2',
    indicatorName: 'DHSC',
    indicatorDefinition:
      'Total number of patients registered with the practice',
    earliestDataPeriod: '1999',
    latestDataPeriod: '1999',
    dataSource: 'Student article',
    lastUpdatedDate: new Date('November 5, 2023'),
    hasInequalities: true,
    unitLabel: '',
  },
];

const MOCK_DATA_LASTUPDATED_INEQUALITIES: IndicatorDocument = {
  indicatorID: '101',
  indicatorName: 'NHS',
  indicatorDefinition: 'Total number of patients registered with the practice',
  earliestDataPeriod: '1999',
  latestDataPeriod: '2023',
  dataSource: 'NHS website',
  lastUpdatedDate: new Date('December 6, 2024'),
  hasInequalities: true,
  unitLabel: '',
};

const mockHandleClick = jest.fn();

const mockSetIsLoading = jest.fn();
const mockLoaderContext: LoaderContext = {
  getIsLoading: jest.fn(),
  setIsLoading: mockSetIsLoading,
};
jest.mock('@/context/LoaderContext', () => {
  return {
    useLoadingState: () => mockLoaderContext,
  };
});

const mockGetSearchState = jest.fn();
const mockSearchStateContext: SearchStateContext = {
  getSearchState: mockGetSearchState,
  setSearchState: jest.fn(),
};
jest.mock('@/context/SearchStateContext', () => {
  return {
    useSearchState: () => mockSearchStateContext,
  };
});

const initialSearchState: SearchStateParams = {
  [SearchParams.SearchedIndicator]: 'test',
  [SearchParams.AreasSelected]: ['Area1'],
};

beforeEach(() => {
  user = userEvent.setup();
  mockGetSearchState.mockReturnValue(initialSearchState);
});

describe('content', () => {
  it('should have search result list item', () => {
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(screen.getByRole('listitem')).toBeInTheDocument();
  });

  it('should contain 2 paragraphs and a heading', () => {
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(screen.getAllByRole('paragraph')).toHaveLength(2);
    expect(screen.getByRole('heading')).toBeInTheDocument();
  });

  it('should contain expected text', () => {
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(screen.getByRole('heading').textContent).toContain('NHS');
    expect(screen.getAllByRole('paragraph').at(0)?.textContent).toContain(
      '2023'
    );
    expect(screen.getAllByRole('paragraph').at(1)?.textContent).toContain(
      '06 December 2024'
    );
  });

  it('should display tag if indicator date is within one month of server date', () => {
    const currentDate = new Date(MOCK_DATA[0].lastUpdatedDate);
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        handleClick={mockHandleClick}
        currentDate={currentDate}
      />
    );
    expect(screen.queryByText('Updated in last month')).toBeInTheDocument();
  });

  it('should not display tag if indicator date is not within one month of server date', () => {
    const currentDate = new Date(MOCK_DATA[0].lastUpdatedDate);
    currentDate.setMonth(currentDate.getMonth() + 1);
    currentDate.setDate(currentDate.getDate() + 1);

    render(
      <SearchResult
        result={MOCK_DATA[0]}
        handleClick={mockHandleClick}
        currentDate={currentDate}
      />
    );

    expect(screen.queryByText('Updated in last month')).not.toBeInTheDocument();
  });

  it('should display tag if inequalities data present for indicator', () => {
    const currentDate = new Date(MOCK_DATA[0].lastUpdatedDate);
    render(
      <SearchResult
        result={MOCK_DATA[1]}
        handleClick={mockHandleClick}
        currentDate={currentDate}
      />
    );
    expect(screen.queryByText('Contains inequality data')).toBeInTheDocument();
  });

  it('should NOT display tag if inequalities data NOT present for indicator', () => {
    const currentDate = new Date(MOCK_DATA[0].lastUpdatedDate);
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        handleClick={mockHandleClick}
        currentDate={currentDate}
      />
    );
    expect(
      screen.queryByText('Contains inequality data')
    ).not.toBeInTheDocument();
  });

  it('should show LAST UPDATE tag then INEQUALITIES tag (left to right) if both present in data', () => {
    const currentDate = new Date(
      MOCK_DATA_LASTUPDATED_INEQUALITIES.lastUpdatedDate
    );
    render(
      <SearchResult
        result={MOCK_DATA_LASTUPDATED_INEQUALITIES}
        handleClick={mockHandleClick}
        currentDate={currentDate}
      />
    );

    const lastUpdatedElement = screen.queryByText('Updated in last month');
    const inequalitiesElement = screen.queryByText('Contains inequality data');

    expect(lastUpdatedElement).toBeInTheDocument();
    expect(inequalitiesElement).toBeInTheDocument();

    if (lastUpdatedElement && inequalitiesElement)
      // IF should always be true but check keeps the various tools happy
      expect(
        lastUpdatedElement.compareDocumentPosition(inequalitiesElement)
      ).toBe(Node.DOCUMENT_POSITION_FOLLOWING);
  });

  it('should display a range of dates if earliest data period and latest data period are different', () => {
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(
      screen.getByText('Data period:', { exact: false })
    ).toHaveTextContent(
      `Data period: ${MOCK_DATA[0].earliestDataPeriod} to ${MOCK_DATA[0].latestDataPeriod}`
    );
  });

  it('should display a single date if earliest data period and latest data period match', () => {
    render(
      <SearchResult result={MOCK_DATA[1]} handleClick={mockHandleClick} />
    );

    expect(
      screen.getByText('Data period:', { exact: false })
    ).toHaveTextContent(`Data period: ${MOCK_DATA[1].earliestDataPeriod}`);
  });
});

describe('Indicator Checkbox', () => {
  it('should mark the checkbox as checked if indicatorSelected is true', () => {
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        indicatorSelected={true}
        handleClick={mockHandleClick}
      />
    );

    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('should mark the checkbox as not checked if indicatorSelected is false', () => {
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        indicatorSelected={false}
        handleClick={mockHandleClick}
      />
    );

    expect(screen.getByRole('checkbox')).not.toBeChecked();
  });

  it('should call the provided handleClick function when checked', async () => {
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        indicatorSelected={false}
        handleClick={mockHandleClick}
      />
    );

    await user.click(screen.getByRole('checkbox'));

    expect(mockHandleClick).toHaveBeenCalledWith('1', true);
  });

  it('should call the provided handleClick function when unchecked', async () => {
    render(
      <SearchResult
        result={MOCK_DATA[0]}
        indicatorSelected={true}
        handleClick={mockHandleClick}
      />
    );

    await user.click(screen.getByRole('checkbox'));

    expect(mockHandleClick).toHaveBeenCalledWith('1', false);
  });

  it('should have a direct link to the indicator chart', () => {
    const expectedPath = `/chart?${SearchParams.SearchedIndicator}=test&${SearchParams.IndicatorsSelected}=${MOCK_DATA[0].indicatorID.toString()}&${SearchParams.AreasSelected}=${initialSearchState[SearchParams.AreasSelected]}`;
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(screen.getByRole('link')).toHaveAttribute('href', expectedPath);
  });

  it('should call setIsLoading with true when clicking on the direct link to the indicator chart', async () => {
    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    const user = userEvent.setup();
    await user.click(screen.getByRole('link'));

    expect(mockSetIsLoading).toHaveBeenCalledWith(true);
  });

  it('should populate the area selected parameter with the code for england if no area is selected', () => {
    const expectedPath = `/chart?${SearchParams.SearchedIndicator}=test&${SearchParams.IndicatorsSelected}=${MOCK_DATA[0].indicatorID.toString()}&${SearchParams.AreasSelected}=${areaCodeForEngland}`;
    const searchState: SearchStateParams = {
      ...initialSearchState,
    };
    searchState[SearchParams.AreasSelected] = undefined;

    mockGetSearchState.mockReturnValue(searchState);

    render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(screen.getByRole('link')).toHaveAttribute('href', expectedPath);
  });

  it('snapshot test', () => {
    const container = render(
      <SearchResult result={MOCK_DATA[0]} handleClick={mockHandleClick} />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });
});
