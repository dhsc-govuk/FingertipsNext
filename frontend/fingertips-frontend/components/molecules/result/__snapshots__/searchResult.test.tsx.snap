// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator Checkbox snapshot test 1`] = `
<DocumentFragment>
  .c4 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  position: relative;
  min-height: 40px;
  padding: 0 0 0 40px;
  clear: left;
  margin-bottom: 10px;
}

.c5 {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  width: 40px;
  height: 40px;
  opacity: 0;
  cursor: pointer;
}

.c5:checked + span:after {
  opacity: 1;
}

.c5:focus + span:before {
  outline: 3px solid transparent;
  outline-offset: 3px;
  box-shadow: 0 0 0 3px #ffdd00;
}

.c5 + span {
  opacity: 1;
  pointer-events: auto;
}

.c6 {
  display: inline-block;
  cursor: pointer;
  padding: 8px 15px 5px;
  -ms-touch-action: manipulation;
  touch-action: manipulation;
}

.c6::before {
  content: '';
  box-sizing: border-box;
  position: absolute;
  top: 0;
  left: 0;
  width: 40px;
  height: 40px;
  border: 2px solid black;
  background: transparent;
}

.c6::after {
  content: '';
  position: absolute;
  top: 11px;
  left: 9px;
  width: 18px;
  height: 7px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  border: solid;
  border-width: 0 0 5px 5px;
  border-top-color: transparent;
  background: transparent;
  opacity: 0;
}

.c0 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 5px;
}

.c7 {
  color: #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  margin-top: 0;
  margin-bottom: 15px;
}

.c8 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c8:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
  background-color: #ffdd00;
}

.c8:link {
  color: #1d70b8;
}

.c8:visited {
  color: #4c2c92;
}

.c8:hover {
  color: #2b8cc4;
}

.c8:active {
  color: #2b8cc4;
}

.c8:focus {
  color: #0b0c0c;
}

.c9 {
  margin: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 15px;
}

.c9 > p {
  margin: 0;
}

.c9 > p > code {
  padding: 0.2em 0.4em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(27,31,35,0.05);
  border-radius: 3px;
}

.c9 > pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f6f8fa;
  border-radius: 3px;
}

.c9 > pre > code {
  display: inline;
  padding: 0;
  margin: 0;
  border: 0;
  overflow: visible;
  line-height: inherit;
  word-wrap: normal;
}

.c3 {
  box-sizing: border-box;
  padding-right: 15px;
  padding-left: 15px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 -15px;
}

.c11 {
  margin: 0;
  border: 0;
  border-bottom: 1px solid #bfc1c3;
}

.c10 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.2;
}

.c2 {
  padding-top: 15px;
  padding-bottom: 0;
}

@media print {
  .c4 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    margin-bottom: 10px;
  }
}

@media print {
  .c0 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c7 {
    color: #000;
  }
}

@media print {
  .c7 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c7 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c7 {
    margin-bottom: 20px;
  }
}

@media print {
  .c8 {
    font-family: sans-serif;
  }
}

@media print {
  .c8[href^="/"]::after,
  .c8[href^="http://"]::after,
  .c8[href^="https://"]::after {
    content: " (" attr(href) ")";
    font-size: 90%;
    word-wrap: break-word;
  }
}

@media print {
  .c9 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c9 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c9 {
    margin-bottom: 20px;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-shrink: 1;
    -ms-flex-negative: 1;
    flex-shrink: 1;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media print {
  .c10 {
    font-size: 14px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c10 {
    font-size: 19px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    padding-top: 15px;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    padding-bottom: 0;
  }
}

<li
    class="c0"
    data-testid="search-result"
  >
    <div
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <label
          class="c4"
        >
          <input
            class="c5"
            data-testid="search-results-indicator-1"
            id="search-results-indicator-1"
            name="indicator"
            type="checkbox"
            value="1"
          />
          <span
            class="c6"
          >
            <h5
              class="c7"
            >
              <a
                class="c8"
                href="/chart?si=test&is=1"
              >
                NHS
              </a>
            </h5>
            <div
              class="c9 c10"
            >
              <p>
                Latest data period: 2023
              </p>
            </div>
            <div
              class="c9 c10"
            >
              <p>
                Data source: NHS website
              </p>
            </div>
            <div
              class="c9 c10"
            >
              <p>
                Last updated: 06 December 2024
              </p>
            </div>
          </span>
        </label>
      </div>
    </div>
    <hr
      class="c11"
    />
  </li>
</DocumentFragment>
`;
